### PR TITLE
Update encryption-hashing.md

### PR DESCRIPTION
**Pull request by hknguyenvu**
Fix encryptedText: Argument of type 'string' is not assignable to parameter of type 'ArrayBufferView'

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Docs
- [ ] Other... Please describe:

## What is the current behavior?
When trying to implement the example code, I received error encryptedText: Argument of type 'string' is not assignable to parameter of type 'ArrayBufferView'

## What is the new behavior?
No error anymore

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by micalevisk** [comment]
I just tested the example myself and it worked as expected with nodejs 16 and TS v1.22.17

```ts
import { createCipheriv, createDecipheriv, randomBytes, scrypt } from 'crypto';
import { promisify } from 'util';

const iv = randomBytes(16);
const password = 'Password used to generate key';

async function main() {
  const key = (await promisify(scrypt)(password, 'salt', 32)) as Buffer;
  const cipher = createCipheriv('aes-256-ctr', key, iv);

  const textToEncrypt = 'Nest';
  const encryptedText = Buffer.concat([
    cipher.update(textToEncrypt),
    cipher.final(),
  ]);

  const decipher = createDecipheriv('aes-256-ctr', key, iv);
  const decryptedText = Buffer.concat([
    decipher.update(encryptedText),
    decipher.final(),
  ]);

  console.log(decryptedText.toString()); // Nest
}

main();
```

**Comment by hknguyenvu** [comment]
@micalevisk Yes, if we test encrypt/decrypt in same function like that, that works as expect.

However I think in fact we will break that function to two functions: encrypt and decrypt with key and iv are stored somewhere so I'm a little bit confused when apply the example in documents. Especially, the decrypt example is written as another file with import {} on top

**Comment by micalevisk** [comment]
I believe as both snippets uses the same variable name (`encryptedText`) it means they are shared, this isn't that clear on the docs but is quite intuitive, isn't?

File: content/security/encryption-hashing.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ Now to decrypt `encryptedText` value:
 ```typescript
 import { createDecipheriv } from 'crypto';
```
Comment (micalevisk) [inline content/security/encryption-hashing.md:35]:
```suggestion
import { createDecipheriv } from 'crypto';
// ... import encryptedText
```

maybe this is the right approach, but I'm unsure.

**Comment by hknguyenvu** [comment]
That's not enough for me. When I does not familiar with `Buffer`, the example take me more than 15 minutes to break the example to separated functions

Maybe this will be better

```typescript
import { createCipheriv, createDecipheriv, randomBytes } from 'crypto';

const iv = randomBytes(16);
const key = randomBytes(32);

const cipher = createCipheriv('aes-256-ctr', key, iv);

const textToEncrypt = 'Nest';
const encryptedText = Buffer.concat([
  cipher.update(textToEncrypt),
  cipher.final(),
]);

let result = encryptedText.toString('hex');

const encryptedInputText = Buffer.from(result, 'hex');
const decipher = createDecipheriv('aes-256-ctr', key, iv);
const decryptedText = Buffer.concat([
  decipher.update(encryptedInputText), 
  decipher.final(),
]);

console.log(decryptedText.toString());

```

File: content/security/encryption-hashing.md [hunk 2]
```diff
@@ -15,30 +15,35 @@ import { createCipheriv, randomBytes, scrypt } from 'crypto';
...
+const password = 'Password used to generate key';
+
 const cipher = createCipheriv('aes-256-ctr', key, iv);
 
 const textToEncrypt = 'Nest';
 const encryptedText = Buffer.concat([
   cipher.update(textToEncrypt),
   cipher.final(),
 ]);
+let encryptedString = encryptedText.toString('hex');
```
Comment (micalevisk) [inline content/security/encryption-hashing.md:31]:
```suggestion
const encryptedString = encryptedText.toString('hex');
```

File: content/security/encryption-hashing.md [hunk 3]
```diff
@@ -15,30 +15,35 @@ import { createCipheriv, randomBytes, scrypt } from 'crypto';
 import { promisify } from 'util';
 
 const iv = randomBytes(16);
-const password = 'Password used to generate key';
 
 // The key length is dependent on the algorithm.
 // In this case for aes256, it is 32 bytes.
-const key = (await promisify(scrypt)(password, 'salt', 32)) as Buffer;
+const key = randomBytes(32);
+const password = 'Password used to generate key';
```
Comment (micalevisk) [inline content/security/encryption-hashing.md:22]: `password` isn't used. I didn't follow why `key` is `randomBytes(32)`

**Comment by micalevisk** [comment]
@hknguyenvu The docs might be not that clear but your changes are quite wrong. Looks like the errors you're getting is due to misusages (we can't tell much as you didn't share any code).

Here's a full working example if you want to split that:


```ts

import { createCipheriv, createDecipheriv, randomBytes, scrypt } from 'crypto';
import { promisify } from 'util';

// another file
function encrypt(textToEncrypt: string, key: Buffer, iv: Buffer): Buffer {
  const cipher = createCipheriv('aes-256-ctr', key, iv);

  const encryptedText = Buffer.concat([
    cipher.update(textToEncrypt),
    cipher.final(),
  ]);

  return encryptedText;
}

// another file
function decrypt(encryptedText: Buffer, key: Buffer, iv: Buffer): string {
  const decipher = createDecipheriv('aes-256-ctr', key, iv);
  const decryptedText = Buffer.concat([
    decipher.update(encryptedText),
    decipher.final(),
  ]);

  return decryptedText.toString();
}


async function main() {
  const iv = randomBytes(16);
  const password = 'Password used to generate key';
  const key = (await promisify(scrypt)(password, 'salt', 32)) as Buffer;

  const textToEncrypt = 'Nest';

  const encryptedText = encrypt(textToEncrypt, key, iv);
  const decryptedTextStr = decrypt(encryptedText, key, iv);
  console.log(decryptedTextStr === textToEncrypt); // true
}
main();
```

refs:

- `Buffer`: https://nodejs.org/api/buffer.html
- `crypto`: https://nodejs.org/api/crypto.html

**Comment by hknguyenvu** [comment]
I think we have many solutions to implement encrypt/decrypt
I don't use a `password`, I will store `key` and `iv` to encrypt/decrypt.

``` typescript
async encryptText(text: string) {
  const key = this.configService.get<string>('aesSecretCode')['key'];
  const iv = this.configService.get<string>('aesSecretCode')['iv'];
  const cipher = createCipheriv('aes-256-ctr', Buffer.from(key), iv);

  const encryptedText = Buffer.concat([
    cipher.update(text),
    cipher.final(),
  ]);

  return encryptedText.toString('hex');
}

async decryptText(text: string) {
  const key = this.configService.get<string>('aesSecretCode')['key'];
  const iv = this.configService.get<string>('aesSecretCode')['iv'];

  const encryptedText = Buffer.from(text, 'hex');
  const decipher = createDecipheriv('aes-256-ctr', key, iv);
  const decryptedText = Buffer.concat([
    decipher.update(encryptedText),
    decipher.final(),
  ]);

  return decryptedText.toString('utf8');
}
```

This design help me to share the encrypted text between multi codebase easier. For example, I have a PHP system and JS system connect to a database, the encrypted text is stored in database. So storing `key` and `iv` help me not to care much about convert PHP code to JS code and vice versa.

> Here's a full working example if you want to split that

If `iv` is a random text in encrypt function, you must store that `iv` somewhere if you save the encrypted text to database? I think that's not good. If you want to just store password, `iv` should depend on the password, something like this:

``` typescript
const iv = password.repeat(16).substr(0, 16)
```

**Comment by kamilmysliwiec** [comment]
This page in the docs serves as a **very basic** example of how you could use hashing/encryption in combination with NestJS. I don't think there's a reason to make it more complicated.

**Pull request closed without merging**